perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -1455,6 +1469,107 @@ private void categorizeMembersForSynchronization(PerunSession sess, List<RichMem
+	 * This method categorize members just for add and remove. This can be done partially just by logins.
+	 *
+	 * This method fill 2 member structures which get as parameters:
+	 * 1. candidateToAdd - New members of the group
+	 * 2. membersToRemove - Former members who are not in synchronized ExtSource now
+	 *
+	 * @param sess
+	 * @param group to be synchronized
+	 * @param subjects list of subjects from extSource (at least login for every subject, can be different by type of extSource)
+	 * @param candidatesToAdd 1. container (more above)
+	 * @param membersToRemove 2. container (more above)
+	 * @param membersSource main source for getting data
+	 * @param source source for getting logins
+	 * @param skippedMembers list of skipped members
+	 * @throws InternalErrorException
+	 * @throws ExtSourceNotExistsException
+	 */
+	private void categorizeMembersForLightweightSynchronization(PerunSession sess, List<RichMember> groupMembers, List<Map<String,String>> subjects, List<Candidate> candidatesToAdd, List<RichMember> membersToRemove, ExtSource membersSource, ExtSource source, List<String> skippedMembers) throws InternalErrorException, ExtSourceNotExistsException {
+		//get all logins from source and create structure with mapping
+		Map<String, Map<String, String>> sourceLogins = new HashMap<>();
+		for (Map<String, String> subject: subjects) {
+			String login = subject.get("login");
+			// Skip subjects, which doesn't have login
+			if (login == null || login.isEmpty()) {
+				log.debug("Subject {} doesn't contain attribute login, skipping.", subject);
+				skippedMembers.add("MemberEntry:[" + subject + "] was skipped because login is missing");
+				continue;
+			}		
+			sourceLogins.put(login, subject);
+		}
+
+		//get all logins of this extSource from the group and also prepare all other members in group without this extSource
+		List<RichMember> richMembersWithoutLogin = new ArrayList<>();
+		Map<String, RichMember> groupLogins = new HashMap<>();
+		for(RichMember rm: groupMembers) {
+			List<UserExtSource> rmues = rm.getUserExtSources();
+			boolean found = false;
+			for(UserExtSource ues: rmues) {
+				if(ues.getExtSource().equals(membersSource)) {
+					groupLogins.put(ues.getLogin(), rm);
+					found = true;
+					break;
+				}
+			}
+			if(!found) richMembersWithoutLogin.add(rm);
+		}
+
+		//prepare special structures of members in group and in source (members in group and source are not interesting for lightweight synchronization)
+		List<RichMember> existInGroupNotInSource = new ArrayList(richMembersWithoutLogin);
+		Map<String, Map<String, String>> existInSourceNotInGroup = new HashMap<>();
+		//found all richMembers who are new from ext source (not in group yet)
+		for(String loginInSource: sourceLogins.keySet()) {
+			if(!groupLogins.containsKey(loginInSource)) {
+				existInSourceNotInGroup.put(loginInSource, sourceLogins.get(loginInSource));
+				groupLogins.remove(loginInSource);
```

Comment:
Condition on line 1525 checks that loginInSource isn't stored in groupLogins, therefore there is no need to remove it.